src/garage/torch/algos/pearl_inference_network.py

```diff
@@ -0,0 +1,224 @@
+# pylint: disable=attribute-defined-outside-init
+"""An inference network that learns a latent probabilistic context variable Z.
+
+It is used in PEARL (Probabilistic Embeddings for Actor-Critic Reinforcement
+Learning). The paper on PEARL can be found at https://arxiv.org/abs/1903.08254.
+Code is adapted from https://github.com/katerakelly/oyster.
+"""
+
+import torch
+from torch import nn as nn
+import torch.nn.functional as F
+
+from garage.torch.utils import np_to_torch
+
+
+class PEARLInferenceNetwork(nn.Module):
+    """A network for infering a posterior over the latent context Z.
+
+    In PEARL, policies are conditioned on current state and a latent context
+    variable Z. This inference network estimates the posterior probability of
+    z given past transitions. It uses context information stored in the
+    encoder to infer the probabilistic value of z and samples from a policy
+    conditioned on z.
+
+    Args:
+        latent_dim (int): Latent context variable dimension.
+        context_encoder (garage.torch.algo.RecurrentEncoder): Recurrent
+            context encoder.
+        policy (garage.torch.policies): Policy used to train the network.
+        **kwargs: Arbitrary keyword arguments including:
+            recurrent (bool): True if encoder is recurrent; false if it is
+                permutation-invariant.
+            use_information_bottlebeck (bool): True if latent context is not
+                deterministic.
+            use_next_obs_in_context (bool): True if next observation is used in
+                distinguishing tasks.
+
+    """
+
+    def __init__(self, latent_dim, context_encoder, policy, **kwargs):
```

Comment:
hmm, there is no need here for **kwargs if your class has no keyword args.

Reply:
@ryanjulian Do you mean that I don't need **kwargs because I don't need it for the super init call?

Reply:
when you add `**kwargs` to the end of a Python function, it will capture any keyword arguments which that function is ever called with, even ones you never expected.

This seems convenient at first (often for writing libraries), but for general-purpose Python code it easily creates hard-to-find bugs.
* Using **kwargs turns off the check for whether you passed a kwarg which wasn't expected.
* Passing **kwargs to a super constructor carries these unexpected arguments all the way down the call chain. This means that if you need to refactor something (e.g. by removing or adding an argument), now you're not sure what is actually getting passed down the call chain.
* When you collect kwargs using **kwargs and pass them to a super constructor, you're hiding which parts of the super constructor you actually use. If you keep your call to the super constructor totally-explicit, it's easy to see the relationship between `super` and the child class
* Swallowing kwargs removes your actual arguments from the call signature. When you document this using the docstring, users will only see `def foo(bar, baz, **kwargs)` -- then they will have to explore the superclass (or 2 or 3) to find the actual arugments. It's better just to copy them and the documentation, even if that seems repetitive, because the user wants to see the documentation for the component they are trying to use as fast as possible, regardless of the internal structure of the library.